docs/langref/relay_expr.rst

```diff
@@ -0,0 +1,462 @@
+employed by NNVM and other dataflow graph-based input formats. The fact that the variables
+are not scoped offers some flexibility in evaluation order compared to :code:`let`
+bindings, though this can also introduce some ambiguity in programs (the
+`developer introduction to the Relay IR`__ includes more detailed discussion
+of this nuance).
+
+__ `Introduction to Relay IR`_
+
+In Relay's text format, a graph binding can be written as below (note the lack of a
+:code:`let` keyword and a semicolon):
+
+.. code-block:: python
+
+   %1 = add(%a, %b)
+   %2 = add(%1, %1)
+   multiply(%2, %2)
+
+Graph bindings are not represented as an AST node in Relay, but rather as meta-variables set
+to reference AST nodes. For example, a program like the above could be constructed in Relay's
+Python front-end by setting *Python variables* equal to the corresponding Relay AST node and
+using the variables repeatedly, as below (a C++ program using the corresponding API bindings
+could accomplish the same thing):
+
+.. code-block:: python
+
+   sum1 = relay.add(a, b)
+   sum2 = relay.add(sum1, sum1)
+   relay.multiply(sum2, sum2)
+
+For development purposes and to enable certain optimizations, Relay includes passes to
+convert between dataflow graphs defined using graph bindings and programs with :code:`let`
+bindings in A-normal form, employed by many compiler optimizations from the functional
+programming community (see `"The Essence of Compiling with Continuations" by
+Flanagan *et al*<https://slang.soe.ucsc.edu/cormac/papers/pldi93.pdf>`__ for a discussion
+of the A-normal form).
+
+=======================
+Control Flow Expression
+=======================
+
+Control flow expressions allow the network topology to change
+based on the value of previously executed expressions.
+
+Call
+~~~~
+
+Expressions with function types in Relay are "callable,"
+meaning that they can be invoked via a function call. These consist of
+any expression that evaluates to a closure (i.e., function expressions
+or global functions) and Relay operators.
+
+When a closure is called, the body is evaluated in the closure's stored
+environment (i.e., using the stored values for free variables) with
+local variable bindings added for each argument; the final value
+obtained by evaluating the body is the call's return value.
+In the case of operators, the implementation is opaque to Relay,
+so the result is left up to the registered TVM implementation.
+
+For example, we can call the previously defined `%fact` because it
+has a function type:
+
+.. code-block:: python
+
+   %fact(10)
+
+A type-polymorphic function can also include type arguments at a call
+site. The type arguments are substituted for type parameters when
+type checking.
+
+Note that all type relations in the function type must hold at each
+call site. Because relations are checked at call sites, this means that
+the relations are checked against the types of the particular arguments
```

Comment:
The relation is checked against the specific types of the arguments to the call. If the call's arguments have types `T_1`, ..., `T_n`, it will be checked if the relation holds on types `T_1`, ..., `T_n`. I'll add an example, as you suggest, to illustrate it.